lua/ttt2/libraries/database.lua

```diff
@@ -0,0 +1,1420 @@
+
+	if oldValue == newValue then return end
+
+	-- Get all callbacks
+	local accessName = dataTable.accessName
+	local cache = callbackCache[accessName]
+
+	if not cache then return end
+
+	-- Call all callbacks
+	local itemNameTable = {itemName, allCallbackString}
+	for i = 1, #itemNameTable do
+		local tempItemName = itemNameTable[i]
+
+		local funcCache = cache[tempItemName]
+
+		if not istable(funcCache) then continue end
+
+		local keyTable = {key, allCallbackString}
+		for j = 1, #keyTable do
+			local tempKey = keyTable[j]
+
+			local funcKeyCache = funcCache[tempKey]
+
+			if not istable(funcKeyCache) then continue end
+
+			-- Execute callbacks
+			for identifier, functions in pairs(funcKeyCache) do
+				for k = 1, #functions do
+					local func = functions[k]
+
+					if not isfunction(func) then continue end
+
+					func(accessName, itemName, key, oldValue, newValue)
+				end
+			end
+		end
+	end
+end
+
+---
+-- Resets the database by returning all values to the default
+-- This is called when a global reset for the given database is done.
+-- @param number index the local index of the database
+-- @realm shared
+-- @internal
+local function ResetDatabase(index)
+	local dataTable = registeredDatabases[index]
+
+	if dataTable and not table.IsEmpty(dataTable.storedData) then
+		for itemName, keys in pairs(dataTable.storedData) do
+			for key in pairs(keys) do
+				OnChange(index, itemName, key, database.GetDefaultValue(dataTable.accessName, itemName, key))
+			end
+		end
+	end
+end
+
+---
+-- Adds a callback to be called when the given sql table entries change
+-- @note itemName and key can both be `nil`. The callback function then gets called on a change of every item or every key
+-- @param string accessName the chosen accessName registered for a given database. HAS NOT TO BE the real database-name! And does not have to be registered yet!
+-- @param[opt] string itemName The name of the item in the database. Leave `nil` if you want a callback for every item
+-- @param[opt] string key The name of the key in the database. Leave `nil` if you want a callback for every key
+-- @param function The callback function(accessName, itemName, key, oldValue, newValue), it's only called if the value actually changed
+-- @param[opt] string identifier An identifier by which you can remove the callback more granular
+-- @realm shared
+function database.AddChangeCallback(accessName, itemName, key, callback, identifier)
+	-- Allow every accessName in case the database is only later registered
+	if not isstring(accessName) or not isfunction(callback) then return end
+
+	-- If no itemName is given, subscribe to changes of all items
+	if not isstring(itemName) then
+		itemName = allCallbackString
+	end
+
+	-- If no key is given, subscribe to changes of all keys of the item	
+	if not isstring(key) then
+		key = allCallbackString
+	end
+
+	-- If no identifier is given just choose one	
+	if not isstring(identifier) then
+		identifier = allCallbackString
```

Comment:
This seems to be wrong, since it does not choose one, but abuses the same string.

Reply:
Just done like for convars. It doesnt need to be globally unique as well.
https://wiki.facepunch.com/gmod/cvars.AddChangeCallback